### PR TITLE
Start multiple clients

### PR DESCRIPTION
**Pull request by ankit-iitb**

### CODE REVIEW

**Review by gz — APPROVED** [review]

**Review by hunhoffe — COMMENTED** [review]
Generally LGTM! I left a few comments that should perhaps be addressed

File: kernel/tests/integration-test.rs [hunk 1]
```diff
@@ -1972,7 +1965,94 @@ fn s03_shmem_exokernel_fs_prop_test() {
     controller.join().unwrap();
     client.join().unwrap();
 
-    let _ignore = remove_file(&filename);
+    let _ignore = remove_file(SHMEM_PATH);
+}
+
+#[cfg(not(feature = "baremetal"))]
+#[test]
+fn s03_shmem_multiinstance() {
```
[the remark this answers is not in the document]
Comment (ankit-iitb) [reply kernel/tests/integration-test.rs]: Renamed `s03_shmem_multiinstance` -> `s03_shmem_exokernel_multiinstance`.

File: lib/rpc/src/client.rs [hunk 2]
```diff
@@ -64,7 +69,7 @@ impl RPCClient for Client {
         self.transport.recv_msg(hdr, data_out)?;
 
         // Check request & client IDs, and also length of received data
-        if hdr.client_id != self.client_id || hdr.req_id != self.req_id {
+        if self.client_id != 0 && hdr.client_id + 1 != self.client_id || hdr.req_id != self.req_id {
```
[the remark this answers is not in the document]
Comment (ankit-iitb) [reply lib/rpc/src/client.rs:76]: Update the comments!

File: kernel/src/arch/x86_64/rackscale/mod.rs [hunk 3]
```diff
@@ -117,12 +121,12 @@ pub(crate) fn register_pid(remote_pid: usize) -> Result<usize, KError> {
...
                         let mut pmap = PID_MAP.write();
                         pmap.try_reserve(1)?;
-                        pmap.try_insert(remote_pid, local_pid)
-                            .map_err(|_e| KError::FileDescForPidAlreadyAdded)?;
                         debug!(
                             "Mapped remote pid {} to local pid {}",
                             remote_pid, local_pid
                         );
+                        pmap.try_insert(remote_pid, local_pid)
+                            .map_err(|_e| KError::FileDescForPidAlreadyAdded)?;
```
[the remark this answers is not in the document]
Comment (ankit-iitb) [reply kernel/src/arch/x86_64/rackscale/mod.rs:129]: For now, each client has a single process, and the PID they request is their unique machine id; hence the tuple is working.
I agree; we might need a hashmap in the future.

File: kernel/tests/integration-test.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (gz) [reply kernel/tests/integration-test.rs]: we can eventually rename it to `s07_` I think it's a good idea as these goes "beyond single machine tests" (but I think it's fine if not for this PR)

**Pull request merged**